Add @asgardeo/auth-express implementation (#10)
asgardeo/asgardeo-auth-express-sdk: README.md

# Asgardeo Auth Express SDK

![Builder](https://github.com/asgardeo/asgardeo-auth-js-sdk/workflows/Builder/badge.svg)
[![Stackoverflow](https://img.shields.io/badge/Ask%20for%20help%20on-Stackoverflow-orange)](https://stackoverflow.com/questions/tagged/wso2is)
[![Join the chat at https://join.slack.com/t/wso2is/shared_invite/enQtNzk0MTI1OTg5NjM1LTllODZiMTYzMmY0YzljYjdhZGExZWVkZDUxOWVjZDJkZGIzNTE1NDllYWFhM2MyOGFjMDlkYzJjODJhOWQ4YjE](https://img.shields.io/badge/Join%20us%20on-Slack-%23e01563.svg)](https://join.slack.com/t/wso2is/shared_invite/enQtNzk0MTI1OTg5NjM1LTllODZiMTYzMmY0YzljYjdhZGExZWVkZDUxOWVjZDJkZGIzNTE1NDllYWFhM2MyOGFjMDlkYzJjODJhOWQ4YjE)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/wso2/product-is/blob/master/LICENSE)
[![Twitter](https://img.shields.io/twitter/follow/wso2.svg?style=social&label=Follow)](https://twitter.com/intent/follow?screen_name=wso2)

> 🚧 &ensp;&ensp;This project is a work in progress. Please do not use this yet!

Reviewer: We can remove this, right?

  - [Getting Started](#getting-started)
  - [Middleware](#middleware)
    - [asgardeoAuth](#asgardeoauth)
      - [Arguments](#arguments)
      - [Example](#example)
      - [Description](#description)
      - [Example Usage](#example-usage)
    - [isAuthenticated](#isauthenticated)
      - [Description](#description-1)
      - [Example Usage](#example-usage-1)
  - [APIs](#apis)
    - [getIDToken](#getIDToken)
    - [getBasicUserInfo](#getBasicUserInfo)
    - [getOIDCServiceEndpoints](#getOIDCServiceEndpoints)
    - [getAccessToken](#getAccessToken)
    - [revokeAccessToken](#revokeAccessToken)
    - [requestCustomGrant](#requestCustomGrant)
    - [updateConfig](#updateConfig)
    - [isSignOutSuccessful](#isSignOutSuccessful)
    - [isSignOutSuccessful](#isSignOutSuccessful) 
  - [Data Storage](#data-storage)
  - [Models](#models)
    - [ExpressClientConfig](#expressclientconfig)
    - [Store](#store)
    - [cookieConfig](#cookieconfig)
  - [Develop](#develop)
    - [Prerequisites](#prerequisites)
    - [Installing Dependencies](#installing-dependencies)
  - [Contribute](#contribute)
    - [Reporting issues](#reporting-issues)
  - [License](#license)

## Introduction

> Asgardeo Auth Express SDK provides the core methods that are needed to implement OIDC authentication in JavaScript/TypeScript based server side apps written with Express Framework. This SDK wraps around the `@asgardeo/auth-node` to provide framework specific functionalities for Express enabling the developers to use OIDC authentication in the applications with a minimum effort.

Reviewer: ```suggestion
Asgardeo Auth Express SDK implements OIDC authentication in JavaScript/TypeScript-based server-side apps written with Express Framework. This SDK wraps around the `@asgardeo/auth-node` to provide framework-specific functionalities for Express enabling the developers to use OIDC authentication in the applications with minimum effort.
```

## Prerequisite

Create an organization in Asgardeo if you don't already have one. The organization name you choose will be referred to as `<org_name>` throughout this documentation.
If you are using [Asgardeo Cloud](https://wso2.com/asgardeo/) as the identity server, create a **Standard-Based Application** in the console.

## Install

Install the library from the npm registry.

```
npm install @asgardeo/auth-express
```

## Getting Started

```javascript
//Import Express
const express = require('express');

// The SDK provides a client middleware that can be used to carry out the authentication.
const { AsgardeoExpressAuth, isAuthenticated } = require('@asgardeo/auth-express');

// Create a config object containing the necessary configurations.
const config = {
  clientID: "<YOUR_CLIENT_ID>",
  clientSecret: "<YOUR_CLIENT_SECRET>",

>   baseUrl: "<YOUR_BASE_URL>",

Reviewer: ```suggestion
  baseUrl: "https://api.asgardeo.io/t/<org_name>",
```
I think this will be better. Can we check against other repos?

Author: In this SPA SDK it has `baseUrl: "<ADD_BASE_URL_HERE>"`
But I don't see any way in the console to get this base URL directly so adding the org name to the URL will be easy for the users.

  appURL: "http://localhost:3000",
  scope: ["openid", "profile"],

>   enableOIDCSessionManagement: true,

Reviewer: ```suggestion
```

};

//Initialize an Express App
const app = express();

//Use the middleware and pass the config object as the argument.
app.use(AsgardeoExpressAuth(config));

//At this point the default /login and /logout routes should be available.
//Users can use these two routes for authentication.

//A regular route
app.get("/", (req, res) => {
    res.status(200).send("Hello World");
});

//A Protected Route
app.get("/protected", isAuthenticated, (req, res) => {
    res.status(200).send("Hello from Protected Route");
});

//Start the express app on PORT 5000
app.listen(3000, () => { console.log(`Server Started at PORT 3000`);});

> 

Reviewer: ```suggestion
```

```
---
## Middleware

### asgardeoAuth
```TypeScript
asgardeoAuth(config: ExpressClientConfig, store?: Store);
```

#### Arguments

1. config: [`ExpressClientConfig`](#ExpressClientConfig)

   This contains the configuration information needed to implement authentication such as the client ID, server origin etc. Additional configuration information that is needed to configure the client, can be passed down from this object (Eg: A custom login path). To learn more about what attributes can be passed into this object, refer to the [`ExpressClientConfig`](#ExpressClientConfig) section.

   #### Example

   ```TypeScript
   const config = {
       clientID: "<YOUR_CLIENT_ID>",
       clientSecret: "<YOUR_CLIENT_SECRET>",
       baseUrl: "<YOUR_BASE_URL>",
       appURL: "http://localhost:3000",
       scope: ["openid", "profile"],

>        enableOIDCSessionManagement: true,

Reviewer: ```suggestion
```

#### Description

The SDK provides a client middleware called `asgardeoAuth` that provides you with the necessary methods to implement authentication.
You can use this middleware to initiate the `AsgardeoAuth` for your application. By default, the SDK implements the `/login` and `/logout` routes so as soon as you use `asgardeoAuth` middleware, the `/login` and `/logout` routes will be available out of the box for the users to authenticate.

_Note: The default `/login` and `/logout` route names can be customized.To learn more, refer to the [`ExpressClientConfig`](#ExpressClientConfig) section._

#### Example Usage

```TypeScript
app.use(asgardeoAuth(config, store));
```
---

### isAuthenticated
```TypeScript
isAuthenticated;
```

#### Description

This middleware function can be used to protect a route. When this function is passed down to a route, it will check if the session cookie exists on the request and if not it will return a    `401` error and if the cookie is there, the request will proceed as usual.

#### Example Usage

```TypeScript
app.get("/protected", isAuthenticated, (req, res) => {
    res.status(200).send("Hello from Protected Route");
});
```
---

> 

Reviewer: ```suggestion
```
